server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java

```diff
@@ -290,116 +272,74 @@ private ActiveRuleDto doInsert(ActiveRuleChange change, RuleActivatorContext con
     dao.insert(dbSession, activeRule);
     for (Map.Entry<String, String> param : change.getParameters().entrySet()) {
       if (param.getValue() != null) {
-        ActiveRuleParamDto paramDto = ActiveRuleParamDto.createFor(context.ruleParamsByKeys().get(param.getKey()));
+        ActiveRuleParamDto paramDto = ActiveRuleParamDto.createFor(rule.getParam(param.getKey()));
         paramDto.setValue(param.getValue());
         dao.insertParam(dbSession, activeRule, paramDto);
       }
     }
     return activeRule;
   }
 
-  private ActiveRuleDto doUpdate(ActiveRuleChange change, RuleActivatorContext context, DbSession dbSession) {
+  private ActiveRuleDto doUpdate(ActiveRuleChange change, RuleActivationContext context, DbSession dbSession) {
+    ActiveRuleWrapper activeRule = context.getActiveRule();
+    if (activeRule == null) {
+      return null;
+    }
     ActiveRuleDao dao = db.activeRuleDao();
-    ActiveRuleDto activeRule = context.activeRule();
-    if (activeRule != null) {
-      String severity = change.getSeverity();
-      if (severity != null) {
-        activeRule.setSeverity(severity);
-      }
-      ActiveRule.Inheritance inheritance = change.getInheritance();
-      if (inheritance != null) {
-        activeRule.setInheritance(inheritance.name());
-      }
-      activeRule.setUpdatedAt(system2.now());
-      dao.update(dbSession, activeRule);
-
-      for (Map.Entry<String, String> param : change.getParameters().entrySet()) {
-        ActiveRuleParamDto activeRuleParamDto = context.activeRuleParamsAsMap().get(param.getKey());
-        if (activeRuleParamDto == null) {
-          // did not exist
-          if (param.getValue() != null) {
-            activeRuleParamDto = ActiveRuleParamDto.createFor(context.ruleParamsByKeys().get(param.getKey()));
-            activeRuleParamDto.setValue(param.getValue());
-            dao.insertParam(dbSession, activeRule, activeRuleParamDto);
-          }
+    String severity = change.getSeverity();
+    if (severity != null) {
+      activeRule.get().setSeverity(severity);
+    }
+    ActiveRule.Inheritance inheritance = change.getInheritance();
+    if (inheritance != null) {
+      activeRule.get().setInheritance(inheritance.name());
+    }
+    activeRule.get().setUpdatedAt(system2.now());
+    dao.update(dbSession, activeRule.get());
+
+    for (Map.Entry<String, String> param : change.getParameters().entrySet()) {
+      ActiveRuleParamDto activeRuleParamDto = activeRule.getParam(param.getKey());
+      if (activeRuleParamDto == null) {
+        // did not exist
+        if (param.getValue() != null) {
+          activeRuleParamDto = ActiveRuleParamDto.createFor(context.getRule().getParam(param.getKey()));
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: "createFor" is marked "javax.annotation.ParametersAreNonnullByDefault" but parameter 1 could be null. ([squid:S2637](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2637))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvP5mckxG3vIth76&open=AWE-qvP5mckxG3vIth76)

[ISSUE_KEY]: AWE-qvP5mckxG3vIth76